### PR TITLE
feat(Structures): added context-aware structures

### LINKED ISSUE #14: Extender System

### PR DESCRIPTION
**Pull request by kyranet**
Fixes #14

### CODE REVIEW

**Comment by bdistin** [comment]
I like where this is going. I think we need to throw an error on `this.set` and use `super.set` everywhere. Also don't we need to export a const instance instead of the bare class?

File: src/client/caching/Extender.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * The context data for Extender.
+ */
+export interface ExtenderStructures {
+	[K: string]: never;
+}
+
+/**
+ * The callback definition.
+ */
+export interface ExtenderCallback<T, R extends T> {
```
Comment (vladfrangu) [inline src/client/caching/Extender.ts:84]: This could really be a type

File: src/client/caching/Extender.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * The exported singleton instance of the {@link Extender} class.
+ */
+export const extender = new Extender();
+
+/**
+ * The context data for Extender.
+ */
+export interface ExtenderStructures {
+	[K: string]: never;
```
Comment (vladfrangu) [inline src/client/caching/Extender.ts:78]: This will need to be removed once we add any structure

Comment (kyranet) [reply src/client/caching/Extender.ts:78]: I know, without that I get errors in all `keyof T`.

Comment (kyranet) [reply src/client/caching/Extender.ts:74]: I added a TODO so it's clear.

**Pull request merged**